docs/source/learn/trace-vs-execution-time.md

```diff
@@ -0,0 +1,175 @@
+device is not started.
+
+## Device Time: Compilation and Execution on the Accelerator
+
+To correctly profile the performance of your code on the XLA device, you must
+ensure that your timing captures the phase where the device actually does the
+work. This involves:
+
+1. Ensure the traced computational graph is sent to the device for execution.
+
+1. Make sure the Python script wait until the XLA device has completed all its
+assigned computations before taking the final timestamp.
+
+### Triggering Execution and Ensuring Completion
+
+Several mechanisms trigger graph execution and/or ensure completion:
+
+1. `torch_xla.sync(wait=True)`: This is the most direct method for benchmarking.
+It ensures all pending XLA operations are launched and, crucially, blocks the
+Python script until the device finishes.
+
+1. `Data Access/Transfer`: Operations like `tensor.cpu()`, `tensor.item()`, or
+printing an XLA tensor require the actual data. To provide it, PyTorch/XLA must
+execute the graph that produces the tensor and wait for its completion.
+
+1. `torch_xla.core.xla_model.mark_step()`: Commonly used in loops, this signals
+to XLA that a logical unit of work (e.g., a training step) is defined and can be
+executed. It triggers execution but doesn't inherently block Python until device
+completion.
+
+1. `torch_xla.core.xla_model.optimizer_step(optimizer)`: In training, this
+usually implies `mark_step` and ensures relevant computations are done.
+
+1. `torch_xla.core.xla_model.unlazy(tensors)`: Blocks until specified tensors
+are materialized.
+
+### Correctly Measuring Device Time (Compilation + Execution)
+
+To accurately measure the time including device computation:
+
+```Python
+# Assume 'a' and 'b' are XLA tensors
+# start_time = time.perf_counter()
+
+result = torch.matmul(a, b)
+
+# CORRECT PROFILING: Explicitly wait for the XLA device to finish.
+torch_xla.sync(wait=True)
+
+# end_time = time.perf_counter()
+# elapsed_time = end_time - start_time
+```
+
+Here, `elapsed_time` would include:
+
+1. The (small) tracing time.
+
+1. XLA compilation time (if it's the first time this graph is seen or if it
+changed).
+
+1. XLA device execution time.
+
+1. Overhead of `torch_xla.sync(wait=True)`.
+
+By performing warm-up iterations (running the target code with
+`torch_xla.sync(wait=True)` before the timed section), you can often isolate the
+steady-state execution time from the initial compilation time for more
+consistent benchmarking.
+
+## Case Study: Correctly Profiling Loops with `mark_step`
+
+A common scenario involves loops, such as in model training, where
+`mark_step` is used. Consider this structure:
+
+```Python
+def run_model_step():
+    #... XLA tensor operations...
+    pass
+
+# start_loop_time = time.perf_counter()
+for step in range(num_steps):
+  run_model_step()  # Operations are traced
+  torch_xla.core.xla_model.mark_step()    # Graph for this step is submitted for execution
```

Comment:
nit: prefer to use `torch_xla.sync()` for all places we're using `mark_step()`

Reply:
Based on https://github.com/pytorch/xla/pull/9133#discussion_r2093767487, my proposal would be to keep the difference exposed?